sealedenumprocessor/src/main/kotlin/com/livefront/sealedenum/internal/SuperInterfaces.kt

```diff
@@ -0,0 +1,124 @@
+        superInterfaces
+    )
+    return superInterfaces
+}
+
+/**
+ * The recursive implementation for [getAllSuperInterfaces].
+ *
+ * This method adds all valid super interfaces to [superInterfaces].
+ */
+@KotlinPoetMetadataPreview
+private tailrec fun ClassInspector.getAllSuperInterfacesImpl(
+    typeSpec: TypeSpec,
+    parentTypeArguments: List<TypeName>,
+    superInterfaces: MutableList<TypeName>
+) {
+    check(typeSpec.typeVariables.size == parentTypeArguments.size)
+
+    // Associate the type variable names with their type arguments from the subclass.
+    val typeVariableNamesToTypeArguments = typeSpec.typeVariables.zip(parentTypeArguments).toMap()
+
+    // Add all implement interfaces to our list, taking care to substitute the type names and throwing away invalid
+    // interfaces
+    superInterfaces += typeSpec.superinterfaces.keys.mapNotNull { superinterface ->
+        when (val adjustedTypeName = substituteTypeName(typeVariableNamesToTypeArguments, superinterface)) {
+            is ClassName -> adjustedTypeName
+            is ParameterizedTypeName -> {
+                // Check if this interface is valid or not
+                if (adjustedTypeName.typeArguments.any { it is WildcardTypeName }) {
+                    null
+                } else {
+                    adjustedTypeName
+                }
+            }
+            else -> null
+        }
+    }
+
+    // Setup the tail recursion
+    when (val nextTypeName = typeSpec.superclass) {
+        is ClassName -> if (nextTypeName != ANY) {
+            val nextTypeSpec = getTypeSpec(nextTypeName) ?: return
+
+            getAllSuperInterfacesImpl(
+                typeSpec = nextTypeSpec,
+                parentTypeArguments = emptyList(),
+                superInterfaces = superInterfaces
+            )
+        }
+        is ParameterizedTypeName -> {
+            val nextTypeSpec = getTypeSpec(nextTypeName.rawType) ?: return
+
+            getAllSuperInterfacesImpl(
+                typeSpec = nextTypeSpec,
+                parentTypeArguments = nextTypeName.typeArguments.map {
+                    if (it is TypeVariableName) {
+                        typeVariableNamesToTypeArguments[it] ?: it
+                    } else {
+                        it
+                    }
+                },
+                superInterfaces = superInterfaces
+            )
+        }
+        else -> Unit
+    }
+}
+
+@KotlinPoetMetadataPreview
+@Suppress("TooGenericExceptionCaught")
+internal fun ClassInspector.getTypeSpec(className: ClassName): TypeSpec? = try {
+    classFor(className).toTypeSpec(this, className)
+} catch (e: Exception) {
+    null
+}
+
+/**
+ * Substitutes all type variable names within [typeName] with their mappings found in [map].
+ *
+ * Any remaining type variable names that have no mapping are replaced with [STAR].
+ */
+private fun substituteTypeName(map: Map<TypeVariableName, TypeName>, typeName: TypeName): TypeName = when (typeName) {
```

Comment:
Doesn't really matter but this seems like it would work fine as an extension on `TypeName`:

```kotlin
private fun TypeName.substituteTypeName(map: Map<TypeVariableName, TypeName>): TypeName = when (this) {
    ...
}
```

Maybe you just wanted to avoid it being unclear whether or not it modified the `TypeName` in place or returned a new one...

Reply:
I like it! `TypeName`s are wonderfully immutable, so there's no concern there. Changed in https://github.com/livefront/sealed-enum/pull/10/commits/5c163ab36be0beebb1b9a685721bbfa92863094a